packages/capnpc-ts/src/index.ts

```diff
@@ -1,66 +1,81 @@
+import * as Bluebird from 'bluebird';
 import * as capnp from 'capnp-ts';
 import * as s from 'capnp-ts/lib/std/schema.capnp';
+
+import {CodeGeneratorContext} from './code-generator-context';
+export {CodeGeneratorContext} from './code-generator-context';
+
 import initTrace from 'debug';
 
 import {loadRequest, writeTsFiles} from './compiler';
 
 const trace = initTrace('capnpc');
 trace('load');
 
-const chunks: Buffer[] = [];
+import * as E from './errors';
+export let errors = E;
 
-export function main() {
+export async function main(): Promise<void> {
+  return run().thenReturn().tapCatch((reason) => {
+    // tslint:disable-next-line:no-console
+    console.error(reason);
+    process.exit(1);
+  });
+}
 
-  process.stdin.on('data', (chunk: Buffer) => {
+export async function run(): Bluebird<CodeGeneratorContext> {
 
-    trace('reading data chunk (%d bytes)', chunk.byteLength);
+  return Bluebird.try(() => {
 
-    chunks.push(chunk);
+    const chunks: Buffer[] = [];
 
-  });
+    process.stdin.on('data', (chunk: Buffer) => {
 
-  process.stdin.on('end', () => {
+      trace('reading data chunk (%d bytes)', chunk.byteLength);
 
-    trace('read complete');
+      chunks.push(chunk);
 
-    const reqBuffer = new Buffer(chunks.reduce((l, chunk) => l + chunk.byteLength, -1));
+    });
 
-    let i = 0;
+    return (new Bluebird<void>((resolve) => {
 
-    chunks.forEach((chunk, j) => {
+      process.stdin.on('end', () => {
 
-      if (j === chunks.length - 1) {
+        trace('read complete');
 
-        // Exclude the EOF byte.
+        resolve();
 
-        chunk.copy(reqBuffer, i, 0, chunk.byteLength - 1);
+      });
 
-      } else {
+    })).then(() => {
 
-        chunk.copy(reqBuffer, i);
+      const reqBuffer = new Buffer(chunks.reduce((l, chunk) => l + chunk.byteLength, 0));
```

Comment:
I think I started the buffer byte length count at -1 because of strange behavior I was observing on OSX. 0 appears to be the correct starting point for the reducer, though, so this line is now correct.

I'll try to repro again later on OSX and come up with a patch that works across platforms if it's actually broken there. On my Arch box this works :100:.